csrc/ir/builder.cpp

```diff
@@ -340,19 +340,33 @@ Val* SimplifyingIrBuilder::addExpr(
     rhs_dtype = getDataType(rhs);
   }
   if (lhs == nullptr) {
-    return IrBuilder::IrBuilder::create<Val>(rhs, rhs_dtype);
+    return IrBuilder::create<Val>(rhs, rhs_dtype);
+  }
+  // simplify x + y - y as x
+  if (lhs->definition() != nullptr && lhs->definition()->isA<BinaryOp>()) {
+    auto binary_op = lhs->definition()->as<BinaryOp>();
+    if (binary_op->getBinaryOpType() == BinaryOpType::Add) {
+      if (binary_op->rhs()->isConst() &&
+          (bool)(binary_op->rhs()->value() == -rhs)) {
```

Comment:
Would it be preferable to replace this with `binary_op->rhs()->sameAs(rhs)` in `SimplifyingIrBuilder::addExpr(Val* lhs, Val* rhs)` below or would that miss some cases like `SimplifyingIrBuilder::addExpr(v, 2)`?

Reply:
Why? `x + y + y` does not equal to `x`?

Reply:
I meant to try and cancel `(x - y) + y = x` when all three are symbolic

Reply:
We can do that, but it is not the case I am optimizing for. I am making `SimplifyingIrBuilder` capable of simplifying `SimplifyingIrBuilder::addExpr(y, -1)` when `y` is `x + 1`, for that case, `rhs` is not symbolic, and we are not using the `SimplifyingIrBuilder::addExpr(Val* lhs, Val* rhs)` overload at all.